### PR TITLE
DOC: Updates to nditer usage instructions

### LINKED ISSUE #12764: Updates to `nditer` docs?
I was looking through the instructions on how to use `flags=['multi_index']` at https://docs.scipy.org/doc/numpy/reference/arrays.nditer.html#tracking-an-index-or-multi-index. The recommended way of doing

```
it = np.nditer(a, flags=['multi_index'], op_flags=['writeonly'])
with it:
    while not it.finished:
        it[0] = it.multi_index[1] - it.multi_index[0]
        it.iternext()
```

seems a bit stilted and unpythonic, given that you can avoid iterating manually at all:

```
it = np.nditer(a, flags=['multi_index'], op_flags=['writeonly'])
with it:
    for x in it:
        x[...] = it.multi_index[1] - it.multi_index[0]
```

This description appears for all versions at least up to 1.16.0.

I agree that the alternative looping notation should be mentioned, perhaps in a section on "Fine Grained Loop Control" or "Alternative Iteration Method" or something similar, but it's not necessary to show the feature as the only way of iterating with an index.

If my suggestion meets with a favorable response, I'd be happy to submit a patch. Or I can do that if you just want to see whatI had in mind.

This issue is based on my answer to https://stackoverflow.com/q/54220875/2988730

Another item that came up in the same answer is that at least in version 1.12.0, `nditer` was not a context manager and did not have a close method. I propose adding some `since 1.1....` tags in a few places to document this explicitly.

### PR DESCRIPTION
**Pull request by madphysicist**
More intuitive use of iterators. Context manager introduction version noted explicitly.

This is still a rough draft, and all advice is welcome.

Fixes #12764

### CODE REVIEW

**Comment by madphysicist** [comment]
@seberg @mhvk I've opened this in response to the issue. Could you take a look when you get a chance?

File: doc/source/reference/arrays.nditer.rst [hunk 1]
```diff
@@ -209,6 +201,47 @@ raise an exception
...
+Alternative Looping and Element Access
+--------------------------------------
+
+To make its properties more readily accessible during iteration,
+:class:`nditer` has an alternative syntax for iterating, which works
+explicitly with the iterator object itself. With this looping
+construct, the current value is accessible by indexing into the
+iterator. Other properties, such as tracked indices remain as before.
+
+The Python interactive interpreter unfortunately prints out the
```
Comment (mhvk) [inline doc/source/reference/arrays.nditer.rst]: I don't understand this paragraph. I know you just copied it from above, but can we just remove it altogether?

File: doc/source/reference/arrays.nditer.rst [hunk 2]
```diff
@@ -115,6 +115,10 @@ context is exited.
     array([[ 0,  2,  4],
            [ 6,  8, 10]])
 
+Prior to version 1.15.0, :class:`nditer` was not a context manager and did not
```
Comment (mhvk) [inline doc/source/reference/arrays.nditer.rst]: Do we need this history? I think it just distracts from the flow of the text. Or at least be explicit why this is here; maybe something like:
```
If you are writing code that needs to support older versions of numpy, note that prior to 1.15, ...
```

**Review by mhvk — COMMENTED** [review]
Overall, looks good. Mostly a suggestion to remove more text!

File: doc/source/reference/arrays.nditer.rst (hunk 1 above)
Comment (madphysicist) [reply doc/source/reference/arrays.nditer.rst]: I had no idea what it was, so I left it just in case. Glad it's not just me and will remove.

File: doc/source/reference/arrays.nditer.rst (hunk 2 above)
Comment (madphysicist) [reply doc/source/reference/arrays.nditer.rst]: I agree that this doesn't need to be right there, but I didn't know where else to put it. I added it in because I ran into issues with the syntax in a couple of places. Perhaps a note admonition with the text you suggest?

Comment (mhvk) [reply doc/source/reference/arrays.nditer.rst]: That sounds good. Or use `.. note::` (which perhaps gives it too much emphasis, but it is an actual "note").

Comment (madphysicist) [reply doc/source/reference/arrays.nditer.rst]: I can do `**Note:** ...` as a compromise.

Comment (madphysicist) [reply doc/source/reference/arrays.nditer.rst]: The particular use case that triggered the comment was IDEOne, which uses version 1.12 or so.

**Comment by madphysicist** [comment]
Ready to go?

**Comment by madphysicist** [comment]
Bump.

File: doc/source/reference/arrays.nditer.rst [hunk 3]
```diff
@@ -156,41 +161,29 @@ element in a computation. For example, you may want to visit the
 elements of an array in memory order, but use a C-order, Fortran-order,
 or multidimensional index to look up values in a different array.
 
-The Python iterator protocol doesn't have a natural way to query these
-additional values from the iterator, so we introduce an alternate syntax
-for iterating with an :class:`nditer`. This syntax explicitly works
-with the iterator object itself, so its properties are readily accessible
```
Comment (mattip) [inline doc/source/reference/arrays.nditer.rst]: Why did you remove the first sentence?

File: doc/source/reference/arrays.nditer.rst [hunk 4]
```diff
@@ -156,41 +161,29 @@ element in a computation. For example, you may want to visit the
...
-    >>> while not it.finished:
-    ...     print("%d <%s>" % (it[0], it.multi_index), end=' ')
-    ...     it.iternext()
+    >>> for x in it:
+    ...     print("%d <%s>" % (x, it.multi_index), end=' ')
     ...
     0 <(0, 0)> 1 <(0, 1)> 2 <(0, 2)> 3 <(1, 0)> 4 <(1, 1)> 5 <(1, 2)>
 
-    >>> it = np.nditer(a, flags=['multi_index'], op_flags=['writeonly'])
-    >>> with it:
```
Comment (mattip) [inline doc/source/reference/arrays.nditer.rst]: Combining lines 189-190 into one makes the line too long. I also like that `with it` make it very clear the iterator can be used as a context manager

File: doc/source/reference/arrays.nditer.rst (hunk 3 above)
Comment (mattip) [reply doc/source/reference/arrays.nditer.rst]: Ahh, this is expanded on in a separate section. OK.

**Comment by mattip** [comment]
one small nit, but this seems like a better read than the original documentation.

File: doc/source/reference/arrays.nditer.rst (hunk 4 above)
Comment (madphysicist) [reply doc/source/reference/arrays.nditer.rst]: The combined line is 71 characters long (+8 for the `    >>> `). I did not get rid of the `with` block. In fact, I would say that using he more standard notation `with x as y:` emphasizes the point even better.

**Comment by mattip** [comment]
Thanks @madphysicist
